pkg/cluster/machine/machine.go

```diff
@@ -43,72 +40,64 @@ func IsNode(machine *clusterv1.Machine) bool {
 
 // FirstMaster scans the provided array of machines and return the first
 // one which is a "Master" or nil if none.
-func FirstMaster(machines []*clusterv1.Machine) *clusterv1.Machine {
-	for _, machine := range machines {
+// Machines and BareMetalMachines must be in the same order
+func FirstMaster(machines []*clusterv1.Machine, bl []*baremetalspecv1.BareMetalMachine) (*clusterv1.Machine, *baremetalspecv1.BareMetalMachine) {
+	// TODO: validate size and ordering of lists
+	for i, machine := range machines {
 		if IsMaster(machine) {
-			return machine
+			return machine, bl[i]
 		}
 	}
-	return nil
-}
-
-// FirstMasterInArray scans the provided array of machines and return the first
-// one which is a "Master" or nil if none.
-func FirstMasterInArray(machines []clusterv1.Machine) *clusterv1.Machine {
-	for _, machine := range machines {
-		if IsMaster(&machine) {
-			return &machine
-		}
-	}
-	return nil
-}
-
-// Config returns the provided machine's configuration.
-func Config(machine *clusterv1.Machine) (*baremetalspecv1.BareMetalMachineProviderSpec, error) {
-	codec, err := baremetalspecv1.NewCodec()
-	if err != nil {
-		return nil, err
-	}
-	machineSpec, err := codec.MachineProviderFromProviderSpec(machine.Spec.ProviderSpec)
-	if err != nil {
-		return nil, apierrors.InvalidMachineConfiguration("Cannot unmarshal machine's providerSpec field: %v", err)
-	}
-	return machineSpec, err
+	return nil, nil
 }
 
 // ParseManifest parses the provided machines manifest file.
-func ParseManifest(file string) ([]*clusterv1.Machine, error) {
+func ParseManifest(file string) (ml []*clusterv1.Machine, bl []*baremetalspecv1.BareMetalMachine, err error) {
 	f, err := os.Open(file)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return Parse(f)
-}
-
-// Parse parses the provided machines io.Reader.
-func Parse(r io.Reader) ([]*clusterv1.Machine, error) {
-	bytes, err := ioutil.ReadAll(r)
+	ml, bl, err = Parse(f)
 	if err != nil {
-		return nil, err
+		err = errors.Wrapf(err, "parsing %q", file)
 	}
+	return
+}
 
-	list := &clusterv1.MachineList{}
-	err = yaml.Unmarshal(bytes, &list)
-	if err != nil {
-		return nil, err
-	}
+// Parse parses the provided machines io.Reader.
+func Parse(r io.ReadCloser) (ml []*clusterv1.Machine, bl []*baremetalspecv1.BareMetalMachine, err error) {
```

Comment:
FYI: I've coded a pretty cool serializer that can do these things in a generic manner, so we can use that later to avoid writing the same code over and over again. But for a future PR.